[Guo] iP

gqsg opened the pull request:
**DukeNote**
“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukeNote frees your mind of having to remember things you need to do, never forgot anything again It's,
- text-based
- easy to learn, even my aunty knows how to use
- ~~FAST~~ SUPER FAST to use
- All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2.double-click it.
3.add your tasks.
4.let it manage your tasks for you 😉

And it is **FREE!**, no hidden charges!

Features:
- [x] Managing To dos
- [x] Managing deadlines
- [x] Managing events
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:
```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

gqsg committed:
Level-1

gqsg committed:
Added Command.java

gqsg committed:
Remove command.java, added Task.java and ToDoList.java, modified some of the logics in chatbot.java

gqsg committed:
Added, Deadlines.Java, Events.Java, ToDos.Java, reNamed toDoList.java to TaskList.java and modified the input logic in ChatBot.Echo() method

gqsg committed:
Add EmptyDescriptionException.java and InvalidInputException.java and modified the echo method in ChatBot.java

gqsg committed:
Add EmptyDescriptionException.java and InvalidInputException.java and modified the echo method in ChatBot.java

gqsg committed:
Supports deleting of task now

gqsg committed:
Added autoated testing

gqsg committed:
Added in the ability to save the tasks in the file now, the tasks will load the data from the file everytime when the bot starts.

gqsg committed:
Forgot to commit FileDataHandler.java

gqsg committed:
Add in the ability to understand the input as a date instead of a string
Add in date command to display all the event/deadlines of a specific date

gqsg committed:
Update the file when a task is deleted, fix the bug where todo command
won't take in the full input

gqsg committed:
Merge branch 'branch-Level-7'

gqsg committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/ChatBot.java
#	src/main/java/Events.java
#	src/main/java/TaskList.java

gqsg committed:
Making the code more OOP

gqsg committed:
Divide classes into packages

gqsg committed:
changed directory

gqsg committed:
Added support for Gradle

gqsg committed:
Add JUnit tests

gqsg committed:
Added header comments for all public methods

gqsg committed:
Tweak the code to comply with a coding standard

gqsg committed:
Merge branch 'branch-A-JavaDoc'

gqsg committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/command/dateCommand.java
#	src/main/java/duke/command/deleteCommand.java
#	src/main/java/duke/command/markCommand.java
#	src/main/java/duke/command/unMarkCommand.java

gqsg committed:
Changes File names

gqsg committed:
Give users a way to find a task by searching for a keyword.

gqsg committed:
Give users a way to find a task by searching for a keyword.

XenonShawn commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * AddCommand class to add task to tasklist.
+ */
+public class AddCommand extends Command{
```
There should be a space after `Command` :)

XenonShawn commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,37 @@
+     *
+     * @param taskNo The task number to be deleted.
+     */
+    public DeleteCommand(int taskNo){
```
Spaces should be placed after parentheses as well!

XenonShawn commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,21 @@
+    /**
+     * @inheritDoc
+     */
+    public abstract boolean isExit();
```
Is there a reason why you made this an abstract method instead of concretizing it and returning `false` most of the time?

XenonShawn commented on src/main/java/duke/exception/EmptyDescriptionException.java:
```diff
@@ -0,0 +1,14 @@
+/**
+ * Empty description exception.
+ */
+public class EmptyDescriptionException extends  DukeException{
```
There should only be one space after Java keywords, eg `extends`

XenonShawn commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,132 @@
+    public static Command parseCommand(String input) throws DukeException{
+        String[] arguments =  input.trim().split(" ", 2);
+        String commandType = arguments[0];
+            if ("bye".equalsIgnoreCase(commandType)) {
+               return exit();
+            } else if ("list".equalsIgnoreCase(commandType)) {
+                return listAllTasks();
+            } else if ("unmark".equalsIgnoreCase(commandType)) {
+                return unMarkTask(arguments);
+            } else if ("mark".equalsIgnoreCase(commandType)) {
+                 return markTask(arguments);
+            } else if ("todo".equalsIgnoreCase(commandType)) {
+               return addToDo(arguments);
+            } else if ("deadline".equalsIgnoreCase(commandType)) {
+               return addDeadline(arguments);
+            } else if ("event".equalsIgnoreCase(commandType)) {
+              return addEvent(arguments);
+            } else if ("date".equalsIgnoreCase(commandType)) {
+               return listTasksOnDate(arguments);
+            } else if ("delete".equalsIgnoreCase(commandType)) {
+               return deleteTask(arguments);
+            } else {
+                throw new InvalidInputException();
+            }
```
There's an unnecessary extra layer of indentation here. On a side note, is there a reason why you used this instead of a switch case? It might help clean up the code a little

gqsg replied:
Thanks for pointing that out, I am using this because I want to make the commands non-case sensitive by using the equalsIgnoreCase function.

XenonShawn commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,104 @@
+package duke.storage;
+
+import java.nio.file.Files;
+import java.io.File;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.Scanner;
```
You may want to order your packages so that they are consistent, eg keep all from the same package together

XenonShawn commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,104 @@
+        boolean directoryExists = java.nio.file.Files.exists(FOLDER_PATH);
+        boolean fileExists = java.nio.file.Files.exists(FILE_PATH);
+        TaskList taskList = new TaskList();
+        if(!directoryExists) {
```
There should be a space after keywords, `if` in this case

XenonShawn commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,165 @@
+            if (this.getTaskType(task).equals("Deadlines")) {
+                Deadline d = (Deadline) task;
+                if (d.getDate().equals(date)) {
+                    System.out.println(d);
```
Any reason why this task of printing isn't in the `Ui` class instead?

XenonShawn commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,132 @@
+        return new DeleteCommand(taskNo);
+    }
+
+    private static Command markTask (String[] arguments) throws EmptyDescriptionException {
+        if (arguments.length == 1) {
+            throw new EmptyDescriptionException();
+        }
+        int taskNo = Integer.parseInt(arguments[1]);
+        return new MarkCommand(taskNo);
+    }
+
+    private static Command unMarkTask (String[] arguments) throws EmptyDescriptionException {
+        if (arguments.length == 1) {
+            throw new EmptyDescriptionException();
+        }
+        int taskNo = Integer.parseInt(arguments[1]);
+        return new UnMarkCommand(taskNo);
+    }
+
+    private static Command listAllTasks ()  {
+        return new ListCommand();
+    }
+
+    private static Command listTasksOnDate (String[] arguments) throws EmptyDescriptionException{
+        if (arguments.length == 1) {
+            throw new EmptyDescriptionException();
+        }
+        LocalDate date = LocalDate.parse(arguments[1]);
+        return new DateCommand(date);
+    }
+
+    private static Command exit ()  {
+        return new ByeCommand();
+    }
+    private static String extractDateByKeyword (String keyword, String text) {
+        String[] args = text.split(keyword);
+        String date = args[1].trim();
+        return date;
+    }
+}
```
Nice way of keeping your main method clean by putting all of the relevant functionality in at the bottom in private methods 👍

XenonShawn replied:
These methods have an unnecessary space after the name of each function and before the parentheses

XenonShawn reviewed (commented):
Overall relatively clean and easy to understand code, with some minor style revisions required. Great work!

gqsg committed:
Level-10 Add a GUI to Duke. Use the JavaFX technology to implement the
GUI.

gqsg committed:
Updated the test classes

gqsg commented:
Thanks @XenonShawn for pointing all my mistakes out, I have edited them accordingly.

gqsg committed:
Add assertions

The application has not implemented assertions.

Assertion is important to identify possible bug in the code.

Add in assertions at Parser, storage and taskList classes to document
important assertions that should hold allows us to identify bug earlier.

gqsg committed:
Merge pull request #2 from NUSDG/branch-A-Assertions

Branch a assertions

gqsg committed:
Improve code quality

Some of the code qualities are not update to standard.

Codes needs to be read, understood by other developer, hence it's
important to keep the codes clean and tidy.

Break down long method into smaller methods and avoid using magic numbers.

gqsg committed:
Merge pull request #3 from NUSDG/branch-A-CodeQuality

Improve code quality

gqsg committed:
Set up Continuous Integration (CI).

gqsg committed:
Add new command to clone a task

User has to create a similar task from scratch even though the
date/description are the same.

Add the ability to clone a task so that the user can easily create new
task based on existing task.

gqsg committed:
Improve UI

The font of the label is not very clear, same visual format for both
user and bot which can be confusing sometimes.

Tweak the background color and font colour of user and bot labels.

gqsg committed:
Improve UI

The font of the label is not very clear, same visual format for both
user and bot which can be confusing sometimes.

Tweak the background color and font colour of user and bot labels.

gqsg committed:
Add personality

gqsg committed:
Improve UI

gqsg committed:
Add user guide

There is no user guide,

Without user guide, it might be confusing for new user to use the app.

Added user guide to explain about different features of the program
including example inputs and outputs.

gqsg committed:
Add Ui.png and modify read.me

gqsg committed:
Update README.md

gqsg committed:
Update README.md

gqsg committed:
Update Ui.png and fix minor bugs

gqsg committed:
Update README.md, change format of add command

gqsg committed:
Update Ui.png

gqsg committed:
Update README.md
